### PR TITLE
feat: improved http limits and dns cache

### LINKED ISSUE #34: Don't die when strn.pl DNS fails to resolve
This is a placeholder for avoiding the incident from today:

```console
$ curl https://orchestrator.strn.pl/nodes/nearby/
curl: (6) Could not resolve host: orchestrator.strn.pl
```

A single DNS name failure effectively broke the entire gateway.

### PR DESCRIPTION
**Pull request by lidel**
This PR

- improves HTTP limits to allow more concurrency per L1 peer and in general
  - tested with `ab` and these new settings also pass the stress test from #41 
- removes TLS cert validation when we retrieve blocks from `strn.pl` because it has no value:
   -  `bifrost-gw` will fetch raw blocks and discard if received data does not match expected CID
   -  caboose should don't care if TLS cert is valid, nor if the name from cert matches DNS name, because we get raw IPs from orchestrator anyway, and faked it in TLS client :see_no_evil: 
   -  we don't want to break gateway if Saturn ever changes the name in their certs and our client starts erroring on TLS handshake
   - tldr: only value of TLS in `strn.pl` is that we can use HTTP/2, nothing more
- adds DNS cache that will not remove stale records if DNS zone suddenly disappears 
  - protects us from catastrophic failures like #34

Closes #34

### CODE REVIEW

File: dns_cache.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+// How often should we check for successful updates to cached entries
+const dnsCacheRefreshInterval = 5 * time.Minute
+
+func init() {
+	// Configure DNS cache to not remove stale records to protect gateway from
+	// catastrophic failures like https://github.com/ipfs/bifrost-gateway/issues/34
+	options := dnscache.ResolverRefreshOptions{}
+	options.ClearUnused = false
+	options.PersistOnFailure = false
```
Comment (willscott) [inline dns_cache.go:22]: i think this should be `PersistOnFailure = true` in order to not remove stale records to protect from failures

File: blockstore_caboose.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+				TLSClientConfig: &tls.Config{
+					// Saturn use TLS in controversial ways, which sooner or
+					// later will force them to switch away to different domain
+					// name and certs, in which case they will break us. Since
+					// we are fetching raw blocks and dont really care about
+					// TLS cert being legitimate, let's disable verification
+					// to save CPU and to avoid catastrophic failure when
+					// Saturn L1s suddenly switch to certs with different DNS name.
+					InsecureSkipVerify: true,
+					// ServerName:         "strn.pl",
```
[the remark this answers is not in the document]
Comment (lidel) [reply blockstore_caboose.go:65]: we are not worried about untrusted nodes or IPs being repurposed for something else. if they can't provide useful response, they will get weighted down

we are worried about  TLS certs being rotated into different DNS name and our HTTP client failing to use L1s due to TLS validation error –  this is a real risk [can elaborate in private why]

File: blockstore_caboose.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+				},
```
[the remark this answers is not in the document]
Comment (lidel) [reply blockstore_caboose.go:66]: @guanzo is it possible to disable redirect from `http://` to `https://` if request is for CAR or a Block? These request types do not need TLS protection.

We would like to run tests and see if paying the cost of TLS gives us any perf benefit at all.

[the remark this answers is not in the document]
Comment (willscott) [reply blockstore_caboose.go:66]: some transport encryption so that a passive network observer can't see what CIDs are being accessed seems still useful

File: dns_cache.go (hunk 1 above)
Comment (lidel) [reply dns_cache.go:22]:
```suggestion
	options.PersistOnFailure = true
```

File: blockstore_caboose.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lidel) [reply blockstore_caboose.go:66]: I dont see how, everything is in the clear text on DHT, and here you get CIDs comigled together across all gateway users. 

Until we have fully privacy-preserving DHT, TLS here is just security theatre.

[the remark this answers is not in the document]
Comment (willscott) [reply blockstore_caboose.go:66]: * these queries are largely going to hit the L1 cache where they won't go back to the DHT - this lets you hide which CIDs are popular at the gateway
* arguing 'we can't improve privacy anywhere until we have it everywhere' doesn't seem like a winning strategy either. Unless this is a _huge_ reduction in performance it doesn't seem like this hurts.

[the remark this answers is not in the document]
Comment (lidel) [reply blockstore_caboose.go:66]: Ack. Even if we don't want to do this, measuring how much it _could buy us_ feels like a useful thing to know.
We can't measure it without disabling redirect I mentioned.  But ok to park this for now.

File: dns_cache.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/rs/dnscache"
+)
+
+// Local DNS cache because in this world things are ephemeral
+var dnsCache = &dnscache.Resolver{}
```
[the remark this answers is not in the document]
Comment (lidel) [reply dns_cache.go]: Thanks, I've refactored things in https://github.com/ipfs/bifrost-gateway/pull/46/commits/9972ef2f695e9bd3cc0fd604e45cee262514aa81, but did not do the blockstore.Close thing.
DNS cache that protects resolution against ephemeral NXDOMAIN is useful beyond caboose,  I've moved it to  main.go.
(Future PRs: pass to namesys, reuse it for DNSLink to protect against censorship and things like [ENS snafus](https://www.coindesk.com/policy/2022/09/19/ethlink-restored-after-ethereum-name-service-wins-injunction-against-godaddy/))

**Review by hacdias — APPROVED** [review]

File: blockstore_caboose.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (guanzo) [reply blockstore_caboose.go:66]: > is it possible to update L1s and disable redirect from http:// to https:// if request is for verifiable CAR or a Block?
doable, ya. Although if the perf difference is significant, maybe this is something you can benchmark locally with a saturn fork?

[the remark this answers is not in the document]
Comment (lidel) [reply blockstore_caboose.go:66]: I won't be able to replicate real world traffic locally, but we have such traffic mirrored  on bifrost-gw-staging env.

[the remark this answers is not in the document]
Comment (Jorropo) [reply blockstore_caboose.go:66]: > some transport encryption so that a passive network observer can't see what CIDs are being accessed seems still useful

The code I sent disable HTTP not TLS, you still do TLS above HTTP1.1.

File: blockstore_caboose.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+				// Saturn Weltschmerz
+				TLSClientConfig: &tls.Config{
+					// Saturn use TLS in controversial ways, which sooner or
+					// later will force them to switch away to different domain
+					// name and certs, in which case they will break us. Since
+					// we are fetching raw blocks and dont really care about
+					// TLS cert being legitimate, let's disable verification
+					// to save CPU and to avoid catastrophic failure when
+					// Saturn L1s suddenly switch to certs with different DNS name.
```
Comment (marten-seemann) [inline blockstore_caboose.go:63]: I'm not sure if this is correct:
1. It doesn't save any significant CPU. All crypto operations are still the same. You're only saving the check if the chain ends in the root CA store, and some trivial string matching on the domain name.
2. Saturn switching to a different cert should be fine, as long as they're doing a redirect. If the plan is to use a cert for domain X on domain Y, well, that's just plain wrong.

Comment (lidel) [reply blockstore_caboose.go:63]: Ack on 1, skipping these checks is just a bonus.  
As for 2,  given last week's incidents, and internal conversations, bifrost-gateway does not take any unnecessary risk with Saturn's TLS/DNS setup.

File: blockstore_caboose.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (guanzo) [reply blockstore_caboose.go:66]: @lidel should be disabled in prod now https://github.com/filecoin-saturn/L1-node/commit/108e0c6d3fd7e6dc1a1b9b9426a76e5900c0883c
